[Cocoa] Defer creating CTFonts a little longer for "special case" fonts

litherum opened the pull request:
#### 99a9366c1abfafef23c42f8f73d5df2508bfcc36
<pre>
[Cocoa] Defer creating CTFonts a little longer for &quot;special case&quot; fonts
<a href="https://bugs.webkit.org/show_bug.cgi?id=252035">https://bugs.webkit.org/show_bug.cgi?id=252035</a>
rdar://105259235

Reviewed by Cameron McCormack.

<a href="https://bugs.webkit.org/show_bug.cgi?id=247987">https://bugs.webkit.org/show_bug.cgi?id=247987</a> is about how we need to defer the
CTFontDescriptor -&gt; CTFont transformation until as late as possible, to avoid
ping-pinging between fonts and descriptors. This pach takes one small step toward
that goal, by preserving the CTFontDescriptor slightly longer when we create
a CTFont for our so-called &quot;special case&quot; fonts.

* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::fontWithFamilySpecialCase):
(WebCore::fontWithFamily):

Canonical link: <a href="https://commits.webkit.org/260149@main">https://commits.webkit.org/260149@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d8244aec50519ba84854dd4dbd361a96ad923470

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107304 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16361 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40101 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116460 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111203 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17786 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7598 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/99461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113069 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/96507 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41037 "layout-tests (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95323 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/28143 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9393 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29485 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10036 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15547 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49065 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7021 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11589 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -967,7 +967,7 @@ void FontCache::platformInvalidate()
 }
 
-static RetainPtr<CTFontRef> fontWithFamilySpecialCase(const AtomString& family, const FontDescription& fontDescription, float size, AllowUserInstalledFonts allowUserInstalledFonts)
+static RetainPtr<CTFontDescriptorRef> fontWithFamilySpecialCase(const AtomString& family, const FontDescription& fontDescription, float size, AllowUserInstalledFonts allowUserInstalledFonts)
```
Do you think this should be called `fontDescriptionWithFamilySpecialCase` or similar? Or do we conflate "font" and "font description" enough that it doesn't matter.

litherum replied:
Oh, yeah, that's a good point. Distinguishing between them is kind of the whole point of this effort I'm doing. I'll rename.

heycam reviewed (approved):
s/ping-pinging/ping-ponging/ and s/pach/patch/ in the commit message.

litherum committed:
[Cocoa] Defer creating CTFonts a little longer for "special case" fonts
https://bugs.webkit.org/show_bug.cgi?id=252035
rdar://105259235

Reviewed by Cameron McCormack.

https://bugs.webkit.org/show_bug.cgi?id=247987 is about how we need to defer the
CTFontDescriptor -> CTFont transformation until as late as possible, to avoid
ping-pinging between fonts and descriptors. This pach takes one small step toward
that goal, by preserving the CTFontDescriptor slightly longer when we create
a CTFont for our so-called "special case" fonts.

* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::fontWithFamilySpecialCase):
(WebCore::fontWithFamily):

Canonical link: https://commits.webkit.org/260149@main
